### ISSUE TITLE
Infinite loop in the normalizer

### ISSUE THREAD

**Issue by magv** [gituliar/fuchsia#5]
Try this:

```python
x,e = var("x epsilon")
m,t = normalize(matrix([[1/x/2,0],[0,0]]),x,e)
```
... the normalizer will loop forever alternating between the initial matrix and the negative of it.

First I thought that the problem is in [these conditions](https://github.com/gituliar/fuchsia/blob/9f66b220479bd241bdff83ef43794cb81c63b7ef/fuchsia/reduction.py#L579):

```python
            balances_x1_x2 = find_balances_helper(a0, b0, x, x1, x2,
                lambda a0_eval, b0_eval: limit_fixed(a0_eval, eps, 0) < 0.5 \
                        and limit_fixed(b0_eval, eps, 0) >= 0.5,
                lambda a0_eval, b0_eval: limit_fixed(a0_eval, eps, 0) >= 0.5 \
                        and limit_fixed(b0_eval, eps, 0) < 0.5,
```

... but changing `< 0.5` into `< -0.5` breaks the normalizer for the other matrices (test suite starts failing). Any ideas?

**Comment by magv** [comment]
OK, I think I know what the problem is: you do need to change `< 0.5` into `< -0.5` (there is a typo in Lee's paper); the test suite is failing after that because algorithm 3 goes like this: if there are points that can be mutually balanced, do that, otherwise try something else; in contrast when `find_balances` finds no pair of point to mutually balance, it gives up.

With `< 0.5` in place, lots of point pairs which are not mutually balanceable where still passing the conditions for such, thus masking the problem in `find_balances`.

* * *

Additionally, `find_balances` never tries to mutually balance points which are not apparent singularities (that is `is_point_fuchsian` is false). This is also a problem.

**Comment by gituliar** [comment]
At first sight it seems for me that it is not possible to normalize the matrix you wrote. The reason is that any balance increases one's point eigenvalue and decreases another's one by 1. See beggining of the Section 6 on p.14:

  > A necessary condition of existance of the eps-form is that eigenvalues of all matrix residues have the form `n+alpha*eps".

What we should do hence is to terminate the loop if this condition does not hold.

**Comment by magv** [comment]
A matrix can be normalizable, but not have an eps-form. These are
separate things generally. It would probably be good to give users the
ability to normalize matrices, even if they won't be able to factor
epsilon later on.

With that said, it is true that this particular matrix can not be
normalized (without introducing additional singularities at least),
because it has two singularities: one at `0` with eigenvalue `1/2` and
one at `oo` with eigenvalue `-1/2`; shifting one of them will just swap
`1/2` and `-1/2` places, thus leaving the matrix unnormalized.

The problem is that current `find_balances` code thinks that this matrix
has valid balances. It has not. This is a symptom of a bigger problem:
`find_balances` does not follow Lee's algorithm 3.

Blacklisting matrices with non-integer eigenvalues will not solve this
problem: it'll just make me come up with another matrix where
`find_balances` misbehaves.

**Comment by magv** [comment]
Also note that algorithm 3 can produce a matrix that is normalized in
all but one point (which is the case for the matrix posted above). The
first paragraph of 6-th section describes how to normalize the remaining
point too (this is done by introducing new singularities). I think we
want that part implemented too.

(This bug report is about a separate problem though).

**Comment by magv** [comment]
Good job with 7f3250c216394fef99132eaca5e0df742246b8be.

BTW, do you remember reference to Bolibrukh's unnormalizable matrix? It turns out that his matrix too has `1/2` and `-1/2` eigenvalues; `normalize` now refuses to normalize it as well.

**Comment by magv** [comment]
Correction: Bolibrukh's matrix does not in fact have `1/2` eigenvalues. It's eigenvalues (after fuchsification) are all in `(-1, 0, 1)`. My mistake here was caused by a bug in `matrix_import`, which is fixed in 325da62b15323c5b22e177416e37228c1455aa12.

In any case, `normalize` refuses to normalize the corrected Bolibrukh's matrix, and I've added a test case for this in 918f74e10fd65e61eeae5f6fc14a6e26981a2d3b.

**Comment by gituliar** [comment]
Concerning non-integer eigenvalues, J.Henn gives an example of such a matrix in [1], eq. (4.13) (you will find this matrix is in the `examples` dir). As we discussed some time ago, balance method does not work in this case, so Henn suggest a transformation of variables which is not rational (BTW, R.Lee also mentions this solution. but with no explicit examples). Maybe it is worth to make some research in this direction and implement it in Fuchsia at some point, e.g. as experimental feature.

[1] http://arxiv.org/abs/1412.2296
